### PR TITLE
feat: GitHub App integration

### PR DESCRIPTION
**Pull request by ryota-sakamoto**
refs: #99

Remaining work:
- [x] write test
- [x] write doc

### CODE REVIEW

**Comment by alexmt** [comment]
@ryota-sakamoto the PR looks awesome! It makes integration with Github much easier. Thank you!

**Comment by ryota-sakamoto** [comment]
@alexmt 
> It makes integration with Github much easier

I think so too, it is very easy.

I'm finished implementation and test, could you review  this PR?

File: pkg/config.go [hunk 1]
```diff
@@ -60,6 +60,34 @@ func replaceStringSecret(val string, secretValues map[string][]byte) string {
 	})
 }
 
+func replaceServiceConfigSecret(data map[string]interface{}, secretValues map[string][]byte) map[string]interface{} {
```
Comment (alexmt) [inline pkg/config.go:63]: Is this change required to make replacement more accurate and avoid accidentally produce invalid YAML? 

The change looks good to me, just want to understand why it was required.

File: docs/services/github.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+![](https://user-images.githubusercontent.com/18019529/108520497-168ce180-730e-11eb-93cb-b0b91f99bdc5.png)
+
+If the message is set to 140 characters or more, it will be truncate.
+
+```yaml
+template.app-deployed: |
+  message: |
+    Application {{.app.metadata.name}} is now running new version of deployments manifests.
+  github:
```
Comment (alexmt) [inline docs/services/github.md:67]: I think we need to leave room for additional actions such as [create commit comment](https://docs.github.com/en/rest/reference/repos#create-a-commit-comment) or [create pull request](https://docs.github.com/en/rest/reference/pulls#create-a-pull-request) (e.g. to create PR that rollback commit if app degrades). To allow adding actions we just need to  move  `state`, `label` and `targetURL` fields under `status` field:

```yaml
github:
  status:  
    state: success
    label: "continuous-delivery/{{.app.metadata.name}}"
    targetURL: "{{.context.argocdUrl}}/applications/{{.app.metadata.name}}?operation=true"
```

What do you think?

**Review by alexmt — CHANGES_REQUESTED** [review]
Sorry, the review took some time since I was on a short vacation. 

PR looks great. Added one proposal to allow extending github integration in the future.

File: pkg/config.go (hunk 1 above)
Comment (ryota-sakamoto) [reply pkg/config.go:63]: That's right.
Currently, `argocd-notifications-secret` is not supported to set multiline value such as private key.
If the user set multiline value, invalid data format of yaml will be created.

File: docs/services/github.md (hunk 2 above)
Comment (ryota-sakamoto) [reply docs/services/github.md:67]: It is good idea.
It can be easily added more actions and `github.status` is more correctly than `github`.

Comment (alexmt) [reply docs/services/github.md:67]: thank you!

**Comment by ryota-sakamoto** [comment]
Hi @alexmt 
I fixed it, and please check it.

**Review by alexmt — APPROVED** [review]
LGTM.

Thank you @ryota-sakamoto !
